modules/core/04-channel/keeper/upgrade_test.go

```diff
@@ -124,4 +128,180 @@ func (suite *KeeperTestSuite) TestChanUpgradeInit() {
+			},
+			connectiontypes.ErrInvalidConnectionState,
+		},
+		{
+			"failed verification for counterparty channel state due to incorrectly constructed counterparty channel",
+			func() {
+				counterpartyChannel.State = types.CLOSED
+			},
+			commitmenttypes.ErrInvalidProof,
+		},
+		{
+			"failed verification for counterparty upgrade due to incorrectly constructed counterparty upgrade",
+			func() {
+				counterpartyUpgrade.LatestSequenceSend = 100
+			},
+			commitmenttypes.ErrInvalidProof,
+		},
+		{
+			"upgrade sequence mismatch, endpointB channel upgrade sequence is ahead",
+			func() {
+				channel := path.EndpointB.GetChannel()
+				channel.UpgradeSequence++
+				path.EndpointB.SetChannel(channel)
+			},
+			types.ErrIncompatibleCounterpartyUpgrade,
+		},
+		{
+			"upgrade ordering is not the same on both sides",
+			func() {
+				upgrade.Fields.Ordering = types.ORDERED
+			},
+			types.ErrIncompatibleCounterpartyUpgrade,
+		},
+		{
+			"proposed connection is not found",
+			func() {
+				upgrade.Fields.ConnectionHops[0] = "invalid-connection"
+			},
+			connectiontypes.ErrConnectionNotFound,
+		},
+		{
+			"proposed connection is not in OPEN state",
+			func() {
+				// reuse existing connection to create a new connection in a non OPEN state
+				connectionEnd := path.EndpointB.GetConnection()
+				connectionEnd.State = connectiontypes.UNINITIALIZED
+				connectionEnd.Counterparty.ConnectionId = counterpartyUpgrade.Fields.ConnectionHops[0] // both sides must be each other's counterparty
+
+				// set proposed connection in state
+				proposedConnectionID := "connection-100"
+				suite.chainB.GetSimApp().GetIBCKeeper().ConnectionKeeper.SetConnection(suite.chainB.GetContext(), proposedConnectionID, connectionEnd)
+				upgrade.Fields.ConnectionHops[0] = proposedConnectionID
+			},
+			connectiontypes.ErrInvalidConnectionState,
+		},
+		{
+			"proposed connection ends are not each other's counterparty",
+			func() {
+				// reuse existing connection to create a new connection in a non OPEN state
+				connectionEnd := path.EndpointB.GetConnection()
+				// ensure counterparty connectionID does not match connectionID set in counterparty proposed upgrade
+				connectionEnd.Counterparty.ConnectionId = "connection-50"
+
+				// set proposed connection in state
+				proposedConnectionID := "connection-100"
+				suite.chainB.GetSimApp().GetIBCKeeper().ConnectionKeeper.SetConnection(suite.chainB.GetContext(), proposedConnectionID, connectionEnd)
+				upgrade.Fields.ConnectionHops[0] = proposedConnectionID
+			},
+			types.ErrIncompatibleCounterpartyUpgrade,
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		suite.Run(tc.name, func() {
+			suite.SetupTest()
+
+			path = ibctesting.NewPath(suite.chainA, suite.chainB)
+			suite.coordinator.Setup(path)
+
+			path.EndpointA.ChannelConfig.ProposedUpgrade.Fields.Version = fmt.Sprintf("%s-v2", mock.Version)
+			err := path.EndpointA.ChanUpgradeInit()
+			suite.Require().NoError(err)
```

Comment:
will open a separate pr with these changes
```diff
diff --git a/testing/config.go b/testing/config.go
index 9a6ace904..90353f3bc 100644
--- a/testing/config.go
+++ b/testing/config.go
@@ -47,9 +47,10 @@ func NewConnectionConfig() *ConnectionConfig {
 }
 
 type ChannelConfig struct {
-       PortID  string
-       Version string
-       Order   channeltypes.Order
+       PortID          string
+       Version         string
+       Order           channeltypes.Order
+       ProposedUpgrade channeltypes.Upgrade
 }
 
 func NewChannelConfig() *ChannelConfig {
diff --git a/testing/endpoint.go b/testing/endpoint.go
index fb278c241..2c82c1581 100644
--- a/testing/endpoint.go
+++ b/testing/endpoint.go
@@ -564,12 +564,30 @@ func (endpoint *Endpoint) TimeoutOnClose(packet channeltypes.Packet) error {
        return endpoint.Chain.sendMsgs(timeoutOnCloseMsg)
 }
 
-func (endpoint *Endpoint) ChanUpgradeInit(timeout channeltypes.Timeout) error {
+func (endpoint *Endpoint) ChanUpgradeInit() error {
+       timeout := endpoint.ChannelConfig.ProposedUpgrade.Timeout
+       if !timeout.IsValid() {
+               timeout = channeltypes.NewUpgradeTimeout(endpoint.Counterparty.Chain.GetTimeoutHeight(), 0)
+       }
+
+       upgradeFields := endpoint.ChannelConfig.ProposedUpgrade.Fields
+       if upgradeFields.Ordering == channeltypes.NONE {
+               upgradeFields.Ordering = endpoint.ChannelConfig.Order
+       }
+
+       if upgradeFields.Version == "" {
+               upgradeFields.Version = endpoint.ChannelConfig.Version
+       }
+
+       if len(upgradeFields.ConnectionHops) == 0 {
+               upgradeFields.ConnectionHops = []string{endpoint.ConnectionID}
+       }
+
        msg := channeltypes.NewMsgChannelUpgradeInit(
                endpoint.ChannelConfig.PortID,
                endpoint.ChannelID,
-               channeltypes.NewUpgradeFields(endpoint.ChannelConfig.Order, []string{endpoint.ConnectionID}, endpoint.ChannelConfig.Version),
-               channeltypes.NewUpgradeTimeout(timeout.Height, timeout.Timestamp),
+               upgradeFields,
+               timeout,
                endpoint.Chain.SenderAccount.GetAddress().String(),
        )
```

Reply:
To clean this code up further, we can add a helper func on endpoint to get the proposed upgrade from the config

Reply:
implemented in #3719